zcash_primitives/src/transaction/builder.rs

```diff
@@ -1215,4 +1438,199 @@ mod tests {
+            1,
+            "There should be only one action"
+        );
+        let action = issuance_builder.get_action(asset).unwrap();
+        assert!(action.is_finalized(), "Action should be finalized");
+        assert_eq!(action.notes().len(), 0, "Action should have zero notes");
+    }
+
+    #[test]
+    #[cfg(zcash_unstable = "nu6" /* TODO nu7 */ )]
+    fn init_issuance_bundle_without_finalization() {
+        let (mut builder, iak, address) = prepare_zsa_test();
+
+        let asset = "asset_desc".to_string();
+
+        builder
+            .init_issuance_bundle::<FeeError>(
+                iak,
+                asset.clone(),
+                Some(IssueInfo {
+                    recipient: address,
+                    value: NoteValue::from_raw(42),
+                }),
+            )
+            .unwrap();
+
+        let issuance_builder = builder.issuance_builder.unwrap();
+        assert_eq!(
+            issuance_builder.actions().len(),
+            1,
+            "There should be only one action"
+        );
+        let action = issuance_builder.get_action(asset).unwrap();
+        assert_eq!(
+            action.is_finalized(),
+            false,
+            "Action should not be finalized"
+        );
+        assert_eq!(action.notes().len(), 1, "Action should have 1 note");
+        assert_eq!(
+            action.notes().first().unwrap().value().inner(),
+            42,
+            "Incorrect note value"
+        );
+    }
+
+    #[test]
+    #[cfg(zcash_unstable = "nu6" /* TODO nu7 */ )]
+    fn add_issuance_same_asset() {
+        let (mut builder, iak, address) = prepare_zsa_test();
+
+        let asset = "asset_desc".to_string();
+
+        builder
+            .init_issuance_bundle::<FeeError>(
+                iak,
+                asset.clone(),
+                Some(IssueInfo {
+                    recipient: address,
+                    value: NoteValue::from_raw(42),
+                }),
+            )
+            .unwrap();
+        builder
+            .add_recipient::<FeeError>(asset.clone(), address, NoteValue::from_raw(21))
+            .unwrap();
+
+        let issuance_builder = builder.issuance_builder.unwrap();
+        assert_eq!(
+            issuance_builder.actions().len(),
+            1,
+            "There should be only one action"
+        );
+        let action = issuance_builder.get_action(asset).unwrap();
+        assert_eq!(
+            action.is_finalized(),
+            false,
+            "Action should not be finalized"
+        );
+        assert_eq!(action.notes().len(), 2, "Action should have 2 notes");
+        assert_eq!(
+            action
+                .notes()
+                .iter()
+                .map(|note| note.value().inner())
+                .sum::<u64>(),
+            42 + 21,
+            "Incorrect notes sum"
+        );
+    }
+
+    #[test]
+    #[cfg(zcash_unstable = "nu6" /* TODO nu7 */ )]
+    fn add_issuance_different_asset() {
```

Comment:
Unfotunately we cannot add burn here and then build because we need correct input to create burn. So we can test it in txtesttool for example. But I've added test for expected failure